### PR TITLE
[Storage] [Typing] [Blob] `_upload_helpers.py` and `_quick_query_helper.py` and `_shared_access_signature.py` and `_list_blob_helpers.py`

### PR DESCRIPTION
**Pull request by vincenttran-msft**
For some reason MyPy keeps flagging this:
azure\storage\blob\_shared_access_signature.py:390: error: Returning Any from function declared to return "str"  [no-any-return]

Which is this line:
```
return sas.generate_account(
        services=Services(blob=True),
        resource_types=resource_types,
```

But even Intellisense infers the return-type correctly??
image

### CODE REVIEW

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 1]
```diff
@@ -93,17 +97,19 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
         # Do single put if the size is smaller than or equal config.max_single_put_size
         if adjusted_count is not None and (adjusted_count <= blob_settings.max_single_put_size):
             try:
-                data = data.read(length)
+                data = data.read(length)  #type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:100]: This whole block of code needs to be reworked but I am unsure how to best do it. The issue is that it currently is "try and if it breaks then we know it wasn't that":
azure\storage\blob\_upload_helpers.py:99: error: Item "bytes" of "Union[bytes, str, Iterable[str], IO[str]]" has no attribute "read"  [union-attr]
azure\storage\blob\_upload_helpers.py:99: error: Item "str" of "Union[bytes, str, Iterable[str], IO[str]]" has no attribute "read"  [union-attr]
azure\storage\blob\_upload_helpers.py:99: error: Item "Iterable[str]" of "Union[bytes, str, Iterable[str], IO[str]]" has no attribute "read"  [union-attr]
azure\storage\blob\_upload_helpers.py:99: error: Item "bytes" of "Union[bytes, str, Iterable[bytes], IO[bytes]]" has no attribute "read"  [union-attr]
azure\storage\blob\_upload_helpers.py:99: error: Item "str" of "Union[bytes, str, Iterable[bytes], IO[bytes]]" has no attribute "read"  [union-attr]
azure\storage\blob\_upload_helpers.py:99: error: Item "Iterable[bytes]" of "Union[bytes, str, Iterable[bytes], IO[bytes]]" has no attribute "read"  [union-attr]
azure\storage\blob\_upload_helpers.py:99: error: Argument 1 to "read" of "IO" has incompatible type "Optional[int]"; expected "int"  [arg-type]

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 2]
```diff
@@ -93,17 +97,19 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
...
                 pass
             if encryption_options.get('key'):
+                if not isinstance(data, bytes):
+                    raise TypeError('Blob data should be of type bytes.')
                 encryption_data, data = encrypt_blob(data, encryption_options['key'], encryption_options['version'])
                 headers['x-ms-meta-encryptiondata'] = encryption_data
 
             response = client.upload(
-                body=data,
+                body=data,  #type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:112]: azure\storage\blob\_upload_helpers.py:112: error: Argument "body" to "upload" of "BlockBlobOperations" has incompatible type "Union[bytes, str, Iterable[str], IO[str]]"; expected "IO[Any]"  [arg-type]
azure\storage\blob\_upload_helpers.py:112: error: Argument "body" to "upload" of "BlockBlobOperations" has incompatible type "Union[bytes, str, Iterable[bytes], 

Again an issue with the typehint of `data`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 3]
```diff
@@ -136,6 +142,10 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
                 cek, iv, encryption_data = generate_blob_encryption_data(
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:142]: azure\storage\blob\_upload_helpers.py:142: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 4]
```diff
@@ -238,7 +249,7 @@ def upload_page_blob(
         response = client.create(
             content_length=0,
             blob_content_length=length,
-            blob_sequence_number=None,
+            blob_sequence_number=None,  #type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:252]: azure\storage\blob\_upload_helpers.py:252: error: Argument "blob_sequence_number" to "create" of "PageBlobOperations" has incompatible type "None"; expected "int"  [arg-type]

How are we passing None here then??

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 5]
```diff
@@ -318,7 +330,7 @@ def upload_append_blob(  # pylint: disable=unused-argument
                 headers=headers,
                 **kwargs)
         except HttpResponseError as error:
-            if error.response.status_code != 404:
+            if error.response.status_code != 404:  #type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:333]: azure\storage\blob\_upload_helpers.py:333: error: Item "None" of "Optional[_HttpResponseCommonAPI]" has no attribute "status_code"  [union-attr]

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py [hunk 6]
```diff
@@ -226,36 +220,29 @@ class BlobPrefix(ItemPaged, DictMixin):
...
+    """The current page of listed results."""
+    delimiter: str
+    """A delimiting character used for hierarchy listing."""
+    command: Callable
+    """Function to retrieve the next page of items."""
+    container: str
+    """The name of the container."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
         super(BlobPrefix, self).__init__(*args, page_iterator_class=BlobPrefixPaged, **kwargs)
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:246]: azure\storage\blob\_list_blobs_helper.py:248: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "str")  [assignment]
azure\storage\blob\_list_blobs_helper.py:249: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "int")  [assignment]
azure\storage\blob\_list_blobs_helper.py:250: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "str")  [assignment]
azure\storage\blob\_list_blobs_helper.py:251: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "str")  [assignment]
azure\storage\blob\_list_blobs_helper.py:252: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "str")  [assignment]

Thanks to having **kwargs, I forgot if we have a solution for this or not 😢

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py [hunk 7]
```diff
@@ -226,36 +222,33 @@ class BlobPrefix(ItemPaged, DictMixin):
...
+    """The location mode being used to list results. The available
+    options include "primary" and "secondary"."""
+    current_page: Optional[List[BlobProperties]]
+    """The current page of listed results."""
+    delimiter: Optional[str]
+    """A delimiting character used for hierarchy listing."""
+    command: Callable
+    """Function to retrieve the next page of items."""
+    container: Optional[str]
+    """The name of the container."""
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:249]: This type is publicly exposed so we may want to think about whether things should actually be typed as Optional here or not. Kind of like we did for `QueueProperties` to try and make the class less annoying to use with MyPy. e.g., `name` likely isn't actually optional?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 6 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:252]: Hmm I feel like this is something we should run into a lot, but I don't remember that actually happening and/or the pattern we use.

Thinking about it more, it's just `Any` so it should just work and I'm not sure why its complaining here. It's probably just the `Optional` part that is actually causing the problem. However, all the member variables are typed as `Any` as well so I'm not sure why it's unhappy.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py (hunk 4 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:250]: Seems like the generated type hint is a bit wrong. Probably because the Swagger is a bit wrong. It has a default value of `0` but `None` is valid and means the request header is just not passed (which is fine). The type ignore is fine here.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 8]
```diff
@@ -254,7 +263,7 @@ def upload_page_blob(
                 kwargs['encryptor'] = encryptor
                 kwargs['padder'] = padder
 
-        kwargs['modified_access_conditions'] = ModifiedAccessConditions(if_match=response['etag'])
+        kwargs['modified_access_conditions'] = ModifiedAccessConditions(if_match=response['etag'])  #type: ignore
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:266]: Error here?

**Review by jalauzon-msft — COMMENTED** [review]
Only reviewed 4 files in PR title.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 6 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:252]: I... don't see any more errors? Must have been a WIP error but it doesn't exist anymore so I'll take it.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py [hunk 9]
```diff
@@ -167,7 +163,7 @@ def seek(self, offset, whence=0):
         if self._point < 0:    # pylint: disable=consider-using-max-builtin
             self._point = 0  # XXX is this right?
 
-    def read(self, size):
+    def read(self, size: int) -> Any:
```
[the remark this answers is not in the document]
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py:166]: I like that idea 😄

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared_access_signature.py [hunk 10]
```diff
@@ -368,30 +387,29 @@ def generate_account_sas(
             :caption: Generating a shared access signature.
     """
     sas = SharedAccessSignature(account_name, account_key)
-    return sas.generate_account(
+    return sas.generate_account(  # type: ignore
```
[the remark this answers is not in the document]
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_shared_access_signature.py:390]: azure\storage\blob\_shared_access_signature.py:390: error: Returning Any from function declared to return "str"  [no-any-return]

Something tells me I need to fix the type hint in the shared SAS file 😢

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py (hunk 8 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:266]: azure\storage\blob\_upload_helpers.py:266: error: Value of type "None" is not indexable  [index]

File: sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py [hunk 11]
```diff
@@ -5,37 +5,36 @@
...
-        error_cls=None,
-    ):
+        name: str = None,  # type: ignore [assignment]
+        container: str = None,  # type: ignore [assignment]
+        errors: Any = None,
+        record_delimiter: str = '\n',
+        encoding: Any = None,
+        headers: Dict[str, Any] = None,  # type: ignore [assignment]
+        response: Any = None,
+        error_cls = None,
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py:36]: Check type for `errors`, `encoding` (should be str), and `error_cls` (probably callable)

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 12]
```diff
@@ -53,7 +56,7 @@ def _convert_mod_error(error):
     raise overwrite_error
 
 
-def _any_conditions(modified_access_conditions=None, **kwargs):  # pylint: disable=unused-argument
+def _any_conditions(modified_access_conditions=None, **kwargs: Any) -> bool:  # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:59]: Or delete if you really want

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 13]
```diff
@@ -63,17 +66,18 @@ def _any_conditions(modified_access_conditions=None, **kwargs):  # pylint: disab
...
-        length=None,
-        overwrite=None,
-        headers=None,
-        validate_content=None,
-        max_concurrency=None,
-        blob_settings=None,
-        encryption_options=None,
-        **kwargs):
+    client: "BlockBlobOperations",
+    data: Union[bytes, str, Iterable[AnyStr], IO[AnyStr]],
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:70]: `data` can't be `str` here.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 14]
```diff
@@ -93,17 +97,19 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
         # Do single put if the size is smaller than or equal config.max_single_put_size
         if adjusted_count is not None and (adjusted_count <= blob_settings.max_single_put_size):
             try:
-                data = data.read(length)
+                data = data.read(length)  # type: ignore
                 if not isinstance(data, bytes):
                     raise TypeError('Blob data should be of type bytes.')
             except AttributeError:
                 pass
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:104]: Replace `try/except` with `hasattr(read)`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py (hunk 8 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py]: Cast it

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 15]
```diff
@@ -146,7 +152,9 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
                     # Adjust total_size for encryption V2
                     total_size = adjusted_count
                     # V2 wraps the data stream with an encryption stream
-                    stream = GCMBlobEncryptionStream(cek, stream)
+                    if cek is None:
+                        raise ValueError("Generate encryption metadata failed. 'cek' is None.")
+                    stream = GCMBlobEncryptionStream(cek, stream)  # type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:157]: `azure\storage\blob\_upload_helpers.py:157: error: Incompatible types in assignment (expression has type "GCMBlobEncryptionStream", variable has type "IO[Any]")  [assignment]`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 16]
```diff
@@ -316,12 +326,12 @@ def upload_append_blob(  # pylint: disable=unused-argument
                 append_position_access_conditions=append_conditions,
                 progress_hook=progress_hook,
                 headers=headers,
-                **kwargs)
+                **kwargs))
         except HttpResponseError as error:
-            if error.response.status_code != 404:
+            if error.response.status_code != 404:  # type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:331]: `azure\storage\blob\_upload_helpers.py:331: error: Item "None" of "Optional[_HttpResponseCommonAPI]" has no attribute "status_code"  [union-attr]`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 17]
```diff
@@ -316,12 +326,12 @@ def upload_append_blob(  # pylint: disable=unused-argument
...
                 raise
             # rewind the request body if it is a stream
-            if hasattr(stream, 'read'):
+            if stream is not None and hasattr(stream, 'read'):
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:334]: Just making a note here that this is why I added this check:
`azure\storage\blob\_upload_helpers.py:337: error: Item "None" of "Optional[IO[Any]]" has no attribute "seek"  [union-attr]`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py (hunk 11 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py:36]: Inside of `query_blob:`
```
errors = kwargs.pop("on_error", None)
error_cls = kwargs.pop("error_cls", BlobQueryError)
encoding = kwargs.pop("encoding", None)
```
And then passed to the `BlobQueryReader` ctor.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 7 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:249]: So I ran `test_list_blobs_with_delimiter` and here are the results:
1st time hitting bp
image

2nd iter
image

3rd iter
image


If I had to guess, first time it enters iteration it's at the root or something so it's `None` and such, but anytime it's operating on a blob that actually exists it seems to be populated? So I am leaning towards type-hinting + throwing the ignores for the _populated_ fields, but want to check with you first

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py [hunk 18]
```diff
@@ -110,49 +106,49 @@ def _extract_data_cb(self, get_next_return):
...
         return self._response.next_marker or None, self.current_page
 
-    def _build_item(self, item):
+    def _build_item(self, item: Union[BlobItemInternal, BlobProperties]) -> BlobProperties:
         if isinstance(item, BlobProperties):
             return item
         if isinstance(item, BlobItemInternal):
             blob = get_blob_properties_from_generated_code(item)  # pylint: disable=protected-access
-            blob.container = self.container
+            blob.container = self.container  # type: ignore [assignment]
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:114]: `azure\storage\blob\_list_blobs_helper.py:114: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 7 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:249]: Look at the BlobPrefix items that the customer gets

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py (hunk 17 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:334]: Make the stream non-optional and then get rid of the none-check

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 7 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:249]: Did you look into what is actually returned to the caller? I would assume name/prefix are always populated? Probably container too? We should type these are non-optional and add ignores.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py [hunk 19]
```diff
@@ -226,74 +222,72 @@ class BlobPrefix(ItemPaged, DictMixin):
...
+    """The name of the container."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
         super(BlobPrefix, self).__init__(*args, page_iterator_class=BlobPrefixPaged, **kwargs)
         self.name = kwargs.get('prefix')
         self.prefix = kwargs.get('prefix')
         self.results_per_page = kwargs.get('results_per_page')
         self.container = kwargs.get('container')
-        self.delimiter = kwargs.get('delimiter')
+        self.delimiter = kwargs.get('delimiter', '/')
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:257]: This seems like a sketchy change. We should either mark delimiter as `Optional` or add an ignore here, depending on if delimiter can ever be `None` when returned.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py [hunk 20]
```diff
@@ -51,7 +53,7 @@ def __init__(
     def __len__(self):
         return self._size
 
-    def _process_record(self, result):
+    def _process_record(self, result: Dict[str, Any]) -> Optional[Any]:
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py:56]: Return type `Optional[bytes]`?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py [hunk 21]
```diff
@@ -67,16 +69,15 @@ def _process_record(self, result):
                 self._errors(error)
         return None
 
-    def _iter_stream(self):
+    def _iter_stream(self) -> Generator[Any, Any, Any]:
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py:72]: `Generator[bytes, None, None]`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py [hunk 22]
```diff
@@ -130,7 +129,7 @@ class QuickQueryStreamer(object):
     File-like streaming iterator.
     """
 
-    def __init__(self, generator):
+    def __init__(self, generator) -> None:
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_quick_query_helper.py:132]: Remove this unless you want to type this whole class. It is not exported though so fine to leave out for now I guess.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared_access_signature.py [hunk 23]
```diff
@@ -299,7 +318,7 @@ def add_resource_signature(self, account_name, account_key, path, user_delegatio
                         sign_string(account_key if user_delegation_key is None else user_delegation_key.value,
                                     string_to_sign))
 
-    def get_token(self):
+    def get_token(self) -> str:
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared_access_signature.py:321]: Again, you only typed one method of many in this class. Either fill in types for the whole class or revert this (since its private).

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 24]
```diff
@@ -63,17 +66,18 @@ def _any_conditions(modified_access_conditions=None, **kwargs):  # pylint: disab
...
+    overwrite: bool = None,  # type: ignore [assignment]
+    encryption_options: Dict[str, Any] = None,  # type: ignore [assignment]
+    blob_settings: "StorageConfiguration" = None,  # type: ignore [assignment]
+    headers: Dict[str, Any] = None,  # type: ignore [assignment]
+    stream: IO = None,  # type: ignore [assignment]
+    length: Optional[int] = None,
+    validate_content: bool = None,  # type: ignore [assignment]
+    max_concurrency: Optional[int] = None,  # type: ignore [assignment]
+    **kwargs: Any
+) -> Dict[str, Any]:
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:80]: Since this is a private method and there is only one caller, go ahead and change this method signature to be non-optional for things that we always pass and remove all ignores. (Like you did for `upload_*_blob` below)

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 25]
```diff
@@ -133,7 +139,7 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
             total_size = length
             encryptor, padder = None, None
             if encryption_options and encryption_options.get('key'):
-                cek, iv, encryption_data = generate_blob_encryption_data(
+                cek, iv, encryption_data = generate_blob_encryption_data(  # type: ignore [assignment]
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:142]: What's this ignore for?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py (hunk 15 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:157]: Can you try changing the type of `data_stream` in the `GCMBlobEncryptionStream` constructor to `IO[bytes]` instead of `BinaryIO` to see if that helps the second error? Will still need to ignore the first but that seems like a good fix for the second.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 7 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:249]: Good call, ran `test_list_blobs_with_delimiter` and should now match.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py (hunk 19 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_list_blobs_helper.py:257]: Hmm yeah so I think my logic here is that a valid `BlobPrefix` is only when a delimiter is provided, and in any instances where we call `walk_blobs`, `/` is provided as a default delimiter. So I think my idea here is that it should be non-None, and if it is None then on the `walk_blobs` side it should have been defaulted to `/`. _BUT_ this logic is wrong in putting it as part of the kwarg pop because that means they specifically passed in `None` (and so while that may cause an exception, I shoudn't be swallowing it and transforming it into a `/` and trying anyways)

tl;dr I had bad logic thinking about this change but now realize what I should do 😄

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py (hunk 25 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:142]: `azure\storage\blob\_upload_helpers.py:142: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]`

[Here](https://github.com/vincenttran-msft/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-blob/azure/storage/blob/_encryption.py#L801-L832) is a link to the relevant API if you think that that means `generate_blob_encryption_data` is mistyping encryption_data as `Optional[str]` (It really shouldn't be optional to my eyes, but we are blocked by `key_encryption_key` making it conditionally set. That can be resolved if we just knew a valid return value (or exception to raise) in the case that `key_encryption_key` is `None`.)

Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py]: I think the issue is that `encryption_data` is used above on line 108 where it gets the type of just `str` and now you are trying to assign it an `Optional[str]`. So, you should be able to solve that somehow... maybe assign a type on line 108, declare the variable higher up as None and assign it a type, or use another variable.

Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py]: You're so right 😄

File: sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py [hunk 26]
```diff
@@ -133,10 +139,10 @@ def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statements
             total_size = length
             encryptor, padder = None, None
             if encryption_options and encryption_options.get('key'):
-                cek, iv, encryption_data = generate_blob_encryption_data(
+                cek, iv, encrypted_data = generate_blob_encryption_data(
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_upload_helpers.py:142]: nit: Could we use a different name than `encrypted_data`. That makes it sound like the variable holds actual data, when it's really just metadata. Maybe `encryption_metadata`?

P.S. I really don't like that MyPy is complaining about this. The first variable is out of scope here so we should be able to use it again (even though scope does not work like that in Python)

**Review by jalauzon-msft — APPROVED** [review]
